### PR TITLE
[doc] Rewrite Quickstart page

### PR DESCRIPTION
**Pull request by ro0gr**
Another attemp to refresh the guides. This time I've focused on the "Quickstart" page only.

In order to explain things I've leveraged a calculator idea which I find a great example of PO usage.

I expect there may be some clumsy turns of speech in the proposed document. Hints to improve are highly appreciated!

My next immediate plan is to refresh a components page, which promises not to be such a huge rewrite like this is.

Part of https://github.com/san650/ember-cli-page-object/issues/412
Partially replaces https://github.com/san650/ember-cli-page-object/pull/415

### CODE REVIEW

File: guides/quickstart.md [hunk 1]
```diff
@@ -4,426 +4,305 @@ title: Quickstart
...
-  enterName: fillable('input.username'),
-  register: clickable('button.register'),
-  myFanfare: MyFanfare
+
+  calculator: Calculator
 });
 ```
 
-Which will allow us to reference the `MyFanfare` component from the acceptance test.
+That's how a minimal application test using a page object can look like:
```
Comment (mistahenry) [inline guides/quickstart.md]: A simple application test using a `page-object` might look like:

**Comment by mistahenry** [comment]
I think the component example with the form would greatly benefit from showing the markup of the form

**Comment by juanazam** [comment]
I agree with all @mistahenry suggestions.

**Comment by ro0gr** [comment]
@mistahenry this is an exact feedback I was looking for. Thanks a lot!

**Comment by ro0gr** [comment]
> I think the component example with the form would greatly benefit from showing the markup of the form

Ya, I've also thought about it. As for me calculator markup takes to much of space on the page:

```html
<form class="quickstart-calculator">
  <span data-test-screen />

  <div class="numpad" >
    <button>1</button>
    <button>2</button>
    <button>3</button>
    <button>4</button>
    <button>5</button>
    <button>6</button>
    <button>7</button>
    <button>8</button>
    <button>9</button>
    <button>0</button>
  </div>

  <button class="plus">+</button>
  <button class="equals">=</button>
</form>
```

I believe we can get rid of such a wasteful template by specifying semantic tags on the PO definition:

```js
export default {
  scope: 'form.quickstart-calculator',

  plus: clickable('button.plus'),

  equals: triggerable('submit'),

  value: value('[data-test-screen]'),

  digits: collection('.numpad > button'),
};
```

Do you still think an explicit template can make documentation better?

---

A possible option might be using of `each` helper, which helps to reduce template size:

```hbs
<form class="quickstart-calculator">
  <span data-test-screen />

  <div class="numpad" >
    {{#each (array 1 2 3 4 5 6 7 8 9 0) as |num|}}
      <button>{{num}}</button>
    {{/each}}
  </div>

  <button class="plus">+</button>
  <button class="equals">=</button>
</form>
```

But then template gets a bit inconsistent cause it lacks the necessary Ember actions to be specified on the buttons and as a consequence the whole calculator js implementation(otherwise it's unclear why do we need actions at all) :)

In other words, until we have something like addon-docs integrated, I think we should avoid handlebars the documentation templates.

File: guides/quickstart.md [hunk 2]
```diff
@@ -4,426 +4,305 @@ title: Quickstart
...
-  beforeEach() {
-    page.setContext(this);
-  },
-  afterEach() {
-    page.removeContext();
-  }
-});
+You also might have noticed that rather than exporting a plain definition we
+export a page object instance which can be used in your tests right after the page
+is imported.
```
[the remark this answers is not in the document]
Comment (ro0gr) [reply guides/quickstart.md]: Yes, this place probably needs an additional explanation. I think a reason for exporting a ready to use page object instances is that we don't need them to be extended.

> for performance?

Not sure what do you mean under performance..

**Comment by mistahenry** [comment]
I really only meant show the form template for the very first component right before this:

```
import { create, triggerable } from 'ember-cli-page-object';

const myForm = create({
  scope: '.my-form',

  datum: { scope: '[data-test-datum]' },

  submit: triggerable('submit')
});
```

I think it's not necessary in the calculator example since the reader could extrapolate from the first example. And as you show above, it's way too much markup to digest in a `quickstart`. But, it may not hurt to casually link to [the dummy app calculator component](https://github.com/san650/ember-cli-page-object/blob/master/tests/dummy/app/templates/calculator.hbs)

File: guides/quickstart.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mistahenry) [reply guides/quickstart.md]: I presumed the reason for advocating exporting page objects rather than definitions was for performance, since creating page object singletons once will be more performant than creating page objects per individual test case.

[the remark this answers is not in the document]
Comment (ro0gr) [reply guides/quickstart.md]: I believe, creating page object instances has very low impact on the test-suite performance. I don't believe we can slowdown a test-suite even if we create page object instances per each test run 😆 

I think that the reason for advocating for exporting of page objects is just a fact that a specific app page has a well known shape(set of components, URL, ...) and it rarely needs to be extended.

File: guides/quickstart.md [hunk 3]
```diff
@@ -4,426 +4,305 @@ title: Quickstart
...
-
-export default create(MyFanfare);
+export default create({
+  visit: visitable('/')
+});
 ```
 
-This separation gives us two `import`-able signatures. In the case of the component's integration test importing the `default` will work as expected:
+The page is provided with a [`visitable`](./api/visitable) attribute to open
+associated web page.
```
[the remark this answers is not in the document]
Comment (ro0gr) [reply guides/quickstart.md]: I'm not sure about the 
> via URL/

when we define a `visitable`, we use a route, like `visitable('/page/:id')`. 

when we use it, we can only pass route params, like:
```
page.visit({id: 1});
```

Going to use your suggestion, but w/o `via URL/` part

[the remark this answers is not in the document]
Comment (mistahenry) [reply guides/quickstart.md]: The main issue was "to open associated web page." must be "to open **the** associated web page".

File: guides/quickstart.md [hunk 4]
```diff
@@ -4,426 +4,278 @@ title: Quickstart
...
 ```
+$ ember generate page-object my-page
 
-And that's it! Our integration and acceptance tests are cleaner, more maintainable and easier to read.
+installing
+  create tests/pages/my-page.js
+```
 
-## Appendix
+The generator created a file inside the directory `/tests/pages`. It allows to easier distinguish pages between components, which are located under `/tests/pages/components/`.
```
Comment (mistahenry) [inline guides/quickstart.md]: "Using this directory allows us to more easily distinguish pages from components, which are located under `/tests/pages/components/`"

**Comment by mistahenry** [comment]
> Going to:
> 
> incorporate the Basics section into the components guide

Having just reread through this page,  I think the calculator example has too much going on as the first thing you'd see when reading a quickstart. The short "Basics" example actually primes you for the calculator example. You explain the concepts more deeply in the components page so I didn't necessarily find the basic section and the components page to be *that* overlapping.

> rename Calculator section to smth like Building a Component
> I believe these changes would move us closer to a real quickstart.

In my opinion, the calculator section should replace the modal example in the components page. This would then be a quickstart with just basics (which is basically components) + pages. Very easy to digest at this point

File: guides/quickstart.md [hunk 5]
```diff
@@ -4,426 +4,172 @@ title: Quickstart
...
-We want to convert these tests to use a page object.
+```html
+<form class="search-form">
+  <input type="search">
+  <button>Search</button>
+</form>
+```
 
-First, we need to create a new page object. For this we'll use one of the generators that comes with the addon.
+Let's generate a page object component definition for it with a help of corresponding generator:
```
Comment (mistahenry) [inline guides/quickstart.md]: Let's generate a page object component definition for it with **the** help of **the** following generator:

File: guides/quickstart.md [hunk 6]
```diff
@@ -4,426 +4,172 @@ title: Quickstart
...
-  visit('/login');
-  fillIn('#username', 'admin');
-  fillIn('#password', 'secret');
-  click('button');
-
-  andThen(function() {
-    assert.equal(currentURL(), '/private-page');
-  });
-});
+The primary build block of page objects is a [component](./components). It consists of a [`scope`](./components#scopes), [`attributes`](./components#attributes), custom methods and nested components.
```
Comment (mistahenry) [inline guides/quickstart.md]: custom methods[,] and nested components.

File: guides/quickstart.md [hunk 7]
```diff
@@ -4,426 +4,172 @@ title: Quickstart
...
+```
 
-  function submit(username, password) {
-    assert.equal(username, 'admin');
-    assert.equal(password, 'secret');
-  }
+## Pages
 
-  this.set('submit', submit);
+For comprehensive testing of our application, we must rely on more than testing individual components in isolation (ie acceptance testing). We can map each "page" in our app (route + template) to a `page-object`, composing various `page-object-component`s together to form a complete representation of page.
```
Comment (mistahenry) [inline guides/quickstart.md]: ...to form a complete representation of **the** page.

File: guides/quickstart.md [hunk 8]
```diff
@@ -4,426 +4,172 @@ title: Quickstart
...
-And that's it! Our integration and acceptance tests are cleaner, more maintainable and easier to read.
+installing
+  create tests/pages/search-page.js
+```
 
-## Appendix
+The generator created a file inside the directory `/tests/pages`. Using this directory allows us to more easily distinguish pages from components, which are located under `/tests/pages/components/`.
 
-A helpful tip is to separate the exports in component page objects. This will allow you to compose larger page objects using the same definitions. For example say we have an integration test of a `my-fanfare` component:
+A generated page output looks like the following:
```
Comment (mistahenry) [inline guides/quickstart.md]: The generated page looks like the following:

**Comment by mistahenry** [comment]
Seems good to go to me.
